include/ginkgo/core/matrix/batch_diagonal.hpp

```diff
@@ -0,0 +1,631 @@
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type* get_const_values() const noexcept
+    {
+        return values_.get_const_data();
+    }
+
+    /**
+     * @copydoc get_values(size_type)
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type* get_const_values(size_type batch) const noexcept
+    {
+        GKO_ASSERT(batch < this->get_num_batch_entries());
+        if (this->get_size().stores_equal_sizes()) {
+            const auto entry_size =
+                std::min(this->get_size().at(0)[0], this->get_size().at(0)[1]);
+            return values_.get_const_data() + batch * entry_size;
+        } else {
+            return values_.get_const_data() +
+                   num_elems_per_batch_cumul_.get_const_data()[batch];
+        }
+    }
+
+    /**
+     * Returns the number of elements explicitly stored in the batch matrix,
+     * cumulative across all the batches.
+     *
+     * @return the number of elements explicitly stored in the matrix,
+     *         cumulative across all the batches
+     */
+    size_type get_num_stored_elements() const noexcept
+    {
+        return values_.get_num_elems();
+    }
+
+    /**
+     * Returns the number of elements explicitly stored at a specific batch
+     * index.
+     *
+     * @param batch  the batch index to be queried
+     *
+     * @return the number of elements explicitly stored in the matrix
+     */
+    size_type get_num_stored_elements(size_type batch) const noexcept
+    {
+        GKO_ASSERT(batch < this->get_num_batch_entries());
+        if (this->get_size().stores_equal_sizes()) {
+            return std::min(this->get_size().at(0)[0],
+                            this->get_size().at(0)[1]);
+        } else {
+            return num_elems_per_batch_cumul_.get_const_data()[batch + 1] -
+                   num_elems_per_batch_cumul_.get_const_data()[batch];
+        }
+    }
+
+    /**
+     * Returns a single element for a particular batch entry.
+     *
+     * @param batch  the batch index to be queried
+     * @param idx  a linear index of the requested element
+     *
+     * @note  the method has to be called on the same Executor the matrix is
+     *        stored at (e.g. trying to call this method on a GPU matrix from
+     *        the OMP results in a runtime error)
+     */
+    ValueType& at(size_type batch, size_type idx) noexcept
+    {
+        return values_.get_data()[linearize_index(batch, idx)];
+    }
+
+    /**
+     * @copydoc BatchDiagonal::at(size_type, size_type, size_type)
+     */
+    ValueType at(size_type batch, size_type idx) const noexcept
+    {
+        return values_.get_const_data()[linearize_index(batch, idx)];
```

Comment:
We should probably add a note this this might not be the most efficient call, though it simplifies the usage and access to data. Unless it is always inlined, the compiler might not be able to optimize kernels using this call.

Reply:
At higher opimization setting, I guess `linearize_index` will be inlined. But still, I'll add a note because in spite of inlining, it might still be slower than manually indexing under specific assumptions known to the application.